connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -117,6 +118,24 @@ private[sql] class AvroDeserializer(
     val incompatibleMsg = errorPrefix +
         s"schema is incompatible (avroType = $avroType, sqlType = ${catalystType.sql})"
 
+    val realDataType = SchemaConverters.toSqlType(avroType).dataType
+    val confKey = SQLConf.LEGACY_AVRO_ALLOW_READING_WITH_INCOMPATIBLE_SCHEMA
+    val preventReadingIncorrectType = !SQLConf.get.getConf(confKey)
+    def incorrectTypeException(provided: DataType): IncompatibleSchemaException = {
+      new IncompatibleSchemaException(errorPrefix + "the original encoded data type is " +
+        s"${realDataType.catalogString}, however you're trying to read the field as " +
+        s"${provided.catalogString}, which would lead to an incorrect answer. To allow reading " +
+        s"this field, enable the SQL configuration: `${confKey.key}`.")
+    }
+
+    def lowerPrecisionException(provided: DataType): IncompatibleSchemaException = {
```

Comment:
Let's move the error to `QueryCompilationErrors` and assign an error class to it.

Reply:
sg!